### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by rjpuca** [jhthorsen/mojo-redis#52]
Hello!
I have a very simple test script that consumes too much memory over time.

```perl
use Mojo::Redis;
$|=1;

my $redis = Mojo::Redis->new( 'redis://localhost:6379' );

sub process_event {
    my $ev = $redis->db->exists( "non_existent_key" );
    print ".";
}

my $i = 0;
while ( 1 ) {
    process_event;
    $i++;
    if ( $i % 1000 == 0 ) {
        print "\n", $i, "\n";
    }
}
```

After about 300000 iterations memory consumption grows from 35 to 350Mb.

(Out of interest I tried to replace Mojo::Redis with other perl redis modules (like Mojo::Redis2 and Redis). Mojo::Redis2 also leaks but to a lesser extent. And with Redis memory consumption stayed the same for almost 500000 iterations, so it obviously doesn't leak.)

**Comment by jhthorsen** [comment]
Thanks for reporting in. I haven't been able to track the issue down, but I started to make some changes in https://github.com/jhthorsen/mojo-redis/compare/jhthorsen/52-memory-leak which seems to be helping a bit. Still unsure what really is going on. Here is the test script I'm using:

```perl
use Mojo::Base -strict;
use Mojo::Redis;
use Test::Memory::Cycle;

my $redis = Mojo::Redis->new('redis://localhost:6379');
my $i = 0;

while (1) {
  $| = 1;
  my $c = $redis->_connection; # Also tried with _blocking_connection()
  $c->on(connect => sub { print '+' });
  $c->on(close => sub { print '-' });
  #memory_cycle_ok($c);
  $c->write_p('EXISTS', 'foo')->then(sub { print $_[0] ? 'e' : 'n' }, sub { print $_[0] })->wait;
  print '.';
  sleep 2 if (++$i) % 3000 == 0;
}
```

Will hopefully find time to look at it more closely during the weekend, but help is greatly appreciated.

**Comment by mohawk2** [comment]
I have had a go at finding what's causing this, using this script modified from both of yours:

```
use Mojo::Base -strict;
use Mojo::Redis;

my $redis = Mojo::Redis->new('redis://localhost');

# these 3 blocks are each different leak-finding snippets
# change the function calls in the if-block to see different ones
# the Devel::Leak is the most readable, but clearly I haven't yet found the problem

my %dump1;
sub dg_start {
require Devel::Gladiator;
%dump1 = map { ("$_" => $_) } @{Devel::Gladiator::walk_arena()};
}
sub dg_end {
#warn Devel::Gladiator::arena_table();
my %dump2 = map { ("$_" => $_) } @{Devel::Gladiator::walk_arena()};
my %dump3 = map { $dump1{$_} ? () : ("$_" => $_) } values %dump2;
warn "found1 ", scalar keys %dump1;
warn "found2 ", scalar keys %dump2;
warn "found3 ", scalar keys %dump3;
#use Devel::Peek; Dump(Mojo::IOLoop->singleton, 20);
require Devel::Peek; Devel::Peek::Dump([ values %dump3 ], 200);
}

my $tracker;
my $gen;
sub deho_start {
require Devel::Cycle;
require Data::Dumper;
require Devel::Events::Handler::ObjectTracker;
require Devel::Events::Filter::RemoveFields;
require Devel::Events::Generator::Objects;
$tracker = Devel::Events::Handler::ObjectTracker->new();
$gen = Devel::Events::Generator::Objects->new(
        handler => Devel::Events::Filter::RemoveFields->new(
                fields => [qw/generator/], # don't need to have a ref to $gen in each event
                handler => $tracker,
        ),
);
$gen->enable(); # start generating events
}
sub deho_end {
$gen->disable();
# live_objects is a Tie::RefHash::Weak hash
my @leaked_objects = keys %{ $tracker->live_objects };
print "leaked ", scalar(@leaked_objects), " objects\n";
foreach my $object ( @leaked_objects ) {
        print "Leaked object: $object\n";
        # the event that generated it
        print Data::Dumper::Dumper( $object, $tracker->live_objects->{$object} );
        Devel::Cycle::find_cycle( $object );
}
}

my $handle;
sub dl_start {
require Devel::Leak;
Devel::Leak::NoteSV($handle);
}
sub dl_end {
Devel::Leak::CheckSV($handle);
}

# if you change the 1 to 0, then the while block will be run, trying to bloat the process
if (1) {
$redis->db;
$redis->db->exists('foo');
$redis->db->exists('foo');
dl_start();
$redis->db->exists('foo');
dl_end();
#undef $redis;
exit;
}

# this is all it takes to make the process grow
while (1) {
  $redis->db->exists('foo');
}
```

I suspect it's due to a closure capturing a strong reference, within a promise. I tried to weaken all the closure-creating code in ::Database, and it started breaking `t/xread.t`. The Devel::Gladiator one does show a cycle. Unfortunately since Devel::Peek follows loops, the output isn't hugely useful yet. Someone who knows the code better may make sense of it!

**Comment by TFBW** [comment]
The leak is in Mojo::Redis::Connection, _parse_message_cb().

$unpack is a CODE ref, and the anonymous sub contains $unpack. Circular reference.

This must be broken by executing "undef $unpack" after calling $unpack at line 216.

There may be other leaks in Mojo::Redis::PubSub. I had one too many issues with that module and banned it from my project.

**Comment by TFBW** [comment]
The _parse_message_cb sub is also missing a check for definedness of $self. Here's a rewrite of that sub which fixes that issue and also eliminates the recursion, so the leak can't happen. I'm using this and haven't encountered any problems so far. You're welcome to incorporate it if you want to.

    sub _parse_message_cb {
      my $self = shift;
    
      Scalar::Util::weaken($self);
      return sub {
        return unless $self;
        my ($protocol, @messages) = @_;
        my $encoding = $self->encoding;
        $self->_write;
    
        my $res = [];
        eval {
          my $in = \@messages;
          my $out = $res;
          while (my $m = shift @$in) {
            if (ref $m eq 'ARRAY') {
              my $subarray = $out;
              ($in, $out) = @$m; # restore saved context
              push @$out, $subarray;
            }
            elsif ($m->{type} eq '-') {
              die "$m->{data}\n";
            }
            elsif ($m->{type} eq ':') {
              push @$out, 0 + $m->{data};
            }
            elsif ($m->{type} eq '*' and ref $m->{data} eq 'ARRAY') {
              my @saved_context = ($in, $out);
              $in = [@{$m->{data}}, \@saved_context];
              $out = [];
            }
            # Only bulk string replies can contain binary-safe encoded data
            elsif ($m->{type} eq '$' and $encoding and defined $m->{data}) {
              push @$out, Mojo::Util::decode($encoding, $m->{data});
            }
            else {
              push @$out, $m->{data};
            }
          }
        };
        my $err = $@;
        chomp $err if $err;
    
        my $p = shift @{$self->{waiting} || []};
        return $p ? $p->reject($err) : $self->emit(error => $err) if $err;
        return $p ? $p->resolve($res->[0]) : $self->emit(response => $res->[0]);
      };
    }

**Comment by hsolomon** [comment]
So solution is this sub along with https://github.com/jhthorsen/mojo-redis/pull/59 ? I need to test as I'm experiencing big memory leak as well. 

Thanks

**Comment by TFBW** [comment]
> So solution is this sub along with #59 ? I need to test as I'm experiencing big memory leak as well.

You can try replacing the version of _parse_message_cb in Mojo::Redis::Connection with the code above. That should address the leak. Or you can use the code from #59 which also fixes the leak. The code above also addresses a minor bug which can result in exceptions when the connection object is destroyed.

**Comment by rabbiveesh** [comment]
Really? I've been getting random exceptions in my tests that use redis. I'm
curious to see if your patch fixes my problem.

**Comment by TFBW** [comment]
Please do share results, and mention whether you are using the PubSub features or not. I have found the patch to be reliable, but I am bypassing the Mojo::Redis::PubSub features in favour of a small module I wrote myself which uses Mojo::Redis::Connection directly.

**Comment by Grinnz** [comment]
> The _parse_message_cb sub is also missing a check for definedness of $self. Here's a rewrite of that sub which fixes that issue and also eliminates the recursion, so the leak can't happen. I'm using this and haven't encountered any problems so far. You're welcome to incorporate it if you want to.

The suggestion is appreciated. I may have a few concerns but it would be easier to analyze them as a PR.

The check for definedness of $self would only avoid an error in the case where the reader is fired after the connection object no longer exists, such as if the Mojo::Redis object itself is discarded - this wouldn't be a real error. If it happens in any other case we should examine that more closely.

**Comment by hsolomon** [comment]
I installed patched release 3.26 from cpan and it is MUCH better.  Memory leak no longer killing my application. Works great. Thank you!
